engine/access/rest/routes/websocket_handler.go

```diff
@@ -0,0 +1,279 @@
+				wsController.wsErrorHandler(models.NewRestError(http.StatusInternalServerError, "failed to set the initial write deadline error: ", err))
+				return
+			}
+			// Write the response to the WebSocket connection
+			err = wsController.conn.WriteJSON(event)
+			if err != nil {
+				wsController.wsErrorHandler(err)
+				return
+			}
+		case <-ticker.C:
+			err := wsController.conn.SetWriteDeadline(time.Now().Add(writeWait))
+			if err != nil {
+				wsController.wsErrorHandler(models.NewRestError(http.StatusInternalServerError, "failed to set the initial write deadline error: ", err))
+				return
+			}
+			if err := wsController.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
+				wsController.wsErrorHandler(err)
+				return
+			}
+		}
+	}
+}
+
+// read function handles WebSocket messages from the client.
+// It continuously reads messages from the WebSocket connection and closes
+// the associated read channel when the connection is closed by client or when an
+// any additional message is received from the client.
+//
+// This method should be called after establishing the WebSocket connection
+// to handle incoming messages asynchronously.
+func (wsController *WebsocketController) read() {
+	// Start a goroutine to handle the WebSocket connection
+	defer close(wsController.readChannel) // notify websocket about closed connection
+
+	for {
+		// reads messages from the WebSocket connection when the connection is closed by client or when an
+		// 1) when the connection is closed by client
+		// 2) when an any additional message is received from the client
+		_, msg, err := wsController.conn.ReadMessage()
+		if err != nil {
+			return
+		}
+
+		// Check the message from the client, if is any just close the connection
+		if len(msg) > 0 {
+			return
+		}
+	}
+}
+
+// SubscribeHandlerFunc is a function that contains endpoint handling logic for subscribes, fetches necessary resources
+type SubscribeHandlerFunc func(
+	ctx context.Context,
+	request *request.Request,
+	wsController *WebsocketController,
+) (state_stream.Subscription, error)
+
+// WSHandler is websocket handler implementing custom websocket handler function and allows easier handling of errors and
+// responses as it wraps functionality for handling error and responses outside of endpoint handling.
+type WSHandler struct {
+	*HttpHandler
+	subscribeFunc SubscribeHandlerFunc
+
+	api               state_stream.API
+	eventFilterConfig state_stream.EventFilterConfig
+	maxStreams        int32
+	streamCount       *atomic.Int32
+}
+
+func NewWSHandler(
+	logger zerolog.Logger,
+	subscribeFunc SubscribeHandlerFunc,
+	chain flow.Chain,
+	api state_stream.API,
+	eventFilterConfig state_stream.EventFilterConfig,
+	maxGlobalStreams uint32,
+) *WSHandler {
+	handler := &WSHandler{
+		subscribeFunc:     subscribeFunc,
+		api:               api,
+		eventFilterConfig: eventFilterConfig,
+		maxStreams:        int32(maxGlobalStreams),
+		streamCount:       &atomic.Int32{},
```

Comment:
```suggestion
		streamCount:       atomic.NewInt32(0),
```

we usually use "go.uber.org/atomic" instead of "sync/atomic"